flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java

```diff
@@ -118,27 +138,162 @@ public static JobExecutionResult submitJobAndWait(
+	 */
+	public static ClassLoader retrieveClassLoader(
+		JobID jobID,
+		ActorGateway jobManager,
+		Configuration config,
+		FiniteDuration timeout)
+		throws JobRetrievalException {
+
+		final Object jmAnswer;
+		try {
+			jmAnswer = Await.result(
+				jobManager.ask(
+					new JobManagerMessages.RequestClassloadingProps(jobID), timeout), timeout);
+		} catch (Exception e) {
+			throw new JobRetrievalException(jobID, "JobManager didn't respond", e);
+		}
+
+		if (jmAnswer instanceof JobManagerMessages.ClassloadingProps) {
+			JobManagerMessages.ClassloadingProps props = ((JobManagerMessages.ClassloadingProps) jmAnswer);
+
+			Option<String> jmHost = jobManager.actor().path().address().host();
+			String jmHostname = jmHost.isDefined() ? jmHost.get() : "localhost";
+			InetSocketAddress serverAddress = new InetSocketAddress(jmHostname, props.blobManagerPort());
+			final BlobCache blobClient = new BlobCache(serverAddress, config);
+
+			final List<BlobKey> requiredJarFiles = props.requiredJarFiles();
+			final List<URL> requiredClasspaths = props.requiredClasspaths();
+
+			final URL[] allURLs = new URL[requiredJarFiles.size() + requiredClasspaths.size()];
+
+			int pos = 0;
+			for (BlobKey blobKey : props.requiredJarFiles()) {
+				try {
+					allURLs[pos++] = blobClient.getURL(blobKey);
+				} catch (Exception e) {
+					blobClient.shutdown();
+					throw new JobRetrievalException(jobID, "Failed to download BlobKey " + blobKey);
+				}
+			}
+
+			for (URL url : requiredClasspaths) {
+				allURLs[pos++] = url;
+			}
+
+			return new URLClassLoader(allURLs, JobClient.class.getClassLoader());
+		} else if (jmAnswer instanceof JobManagerMessages.JobNotFound) {
+			throw new JobRetrievalException(jobID, "Couldn't retrieve class loader. Job " + jobID + " not found");
+		} else {
+			throw new JobRetrievalException(jobID, "Unknown response from JobManager: " + jmAnswer);
+		}
+	}
+
+	/**
+	 * Given a JobListeningContext, awaits the result of the job execution that this context is bound to
+	 * @param listeningContext The listening context of the job execution
+	 * @return The result of the execution
+	 * @throws JobExecutionException if anything goes wrong while monitoring the job
+	 */
+	public static JobExecutionResult awaitJobResult(JobListeningContext listeningContext) throws JobExecutionException {
+
+		final JobID jobID = listeningContext.jobID;
+		final Future<Object> jobSubmissionFuture = listeningContext.jobResultFuture;
+		final ClassLoader classLoader = listeningContext.classLoader;
+
 		// first block handles errors while waiting for the result
-		Object answer;
+		final Object answer;
 		try {
-			Future<Object> future = Patterns.ask(jobClientActor,
-					new JobClientMessages.SubmitJobAndWait(jobGraph),
-					new Timeout(AkkaUtils.INF_TIMEOUT()));
-			
-			answer = Await.result(future, AkkaUtils.INF_TIMEOUT());
+			answer = Await.result(jobSubmissionFuture, AkkaUtils.INF_TIMEOUT());
```

Comment:
What if the `JobClientActor` dies for some reason? Then he won't be able to complete or fail the future and we would be stuck. Maybe we could periodically check if the `JobClientActor` is still alive to avoid this scenario.